pkg/networkserver/downlink.go

```diff
@@ -519,6 +546,76 @@ func appendRecentDownlink(recent []*ttnpb.DownlinkMessage, down *ttnpb.DownlinkM
 	return recent
 }
 
+// downlinkRxDelay returns the RxDelay value, which should be used to answer up.
+func downlinkRxDelay(phy band.Band, macState *ttnpb.MACState, up *ttnpb.UplinkMessage) (del ttnpb.RxDelay) {
+	if up.Payload.MHDR.MType == ttnpb.MType_JOIN_REQUEST {
+		del = ttnpb.RxDelay(phy.JoinAcceptDelay1 / time.Second)
+	} else {
+		del = macState.CurrentParameters.Rx1Delay
+	}
+	if del == ttnpb.RX_DELAY_0 {
+		return ttnpb.RX_DELAY_1
+	}
+	return del
+}
+
+// txRequestFromUplink return the Class A TxRequest, which can be used to answer up.
+// txRequestFromUplink uses downlinkRxDelay to determine RxDelay.
+// txRequestFromUplink does not set the priority.
+func txRequestFromUplink(phy band.Band, macState *ttnpb.MACState, rx1, rx2 bool, up *ttnpb.UplinkMessage) (*ttnpb.TxRequest, error) {
+	if !rx1 && !rx2 {
+		return nil, errNoPath
+	}
+	req := &ttnpb.TxRequest{
+		Class:    ttnpb.CLASS_A,
+		Rx1Delay: downlinkRxDelay(phy, macState, up),
+	}
+	if rx1 {
+		if up.DeviceChannelIndex > math.MaxUint8 {
+			return nil, errInvalidChannelIndex
+		}
+		rx1ChIdx, err := phy.Rx1Channel(uint8(up.DeviceChannelIndex))
+		if err != nil {
+			return nil, err
+		}
+		if uint(rx1ChIdx) >= uint(len(macState.CurrentParameters.Channels)) ||
+			macState.CurrentParameters.Channels[int(rx1ChIdx)] == nil ||
+			macState.CurrentParameters.Channels[int(rx1ChIdx)].DownlinkFrequency == 0 {
+			return nil, errCorruptedMACState
+		}
+		rx1DRIdx, err := phy.Rx1DataRate(up.Settings.DataRateIndex, macState.CurrentParameters.Rx1DataRateOffset, macState.CurrentParameters.DownlinkDwellTime)
+		if err != nil {
+			return nil, err
+		}
+		req.Rx1DataRateIndex = rx1DRIdx
+		req.Rx1Frequency = macState.CurrentParameters.Channels[int(rx1ChIdx)].DownlinkFrequency
+	}
+	if rx2 {
+		req.Rx2DataRateIndex = macState.CurrentParameters.Rx2DataRateIndex
+		req.Rx2Frequency = macState.CurrentParameters.Rx2Frequency
+	}
+	return req, nil
+}
+
+// maximumUplinkLength returns the maximum length of the next uplink after ups.
+func maximumUplinkLength(fp *frequencyplans.FrequencyPlan, phy band.Band, ups ...*ttnpb.UplinkMessage) uint16 {
+	// NOTE: If no data uplink is found, we assume ADR is off on the device and, hence, data rate index 0 is used in computation.
+	maxUpDRIdx := ttnpb.DATA_RATE_0
+loop:
+	for i := len(ups) - 1; i >= 0; i-- {
+		switch ups[i].Payload.MHDR.MType {
+		case ttnpb.MType_JOIN_REQUEST:
```

Comment:
Idem

Reply:
Rejoin-request does not indicate the reset of ADR parameters, so it should be skipped as it's done now